subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	uint8_t current_data_offset;
+};
+
+void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+/* Read data from the flattened AD stream. */
+uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
+{
+	const struct bt_data *current_data = &stream->ad[stream->ad_index].data[stream->data_index];
+
+	const bool is_first_data = (stream->ad_index == 0) && (stream->data_index == 0);
+	const bool is_last_data = (stream->ad_index == stream->ad_len - 1) &&
+				  (stream->data_index == stream->ad[stream->ad_index].len - 1);
```

Comment:
Given the multiple uses of `stream->ad_index` and `stream->data_index` I would suggest to extract these as `const` variables and use those instead of deferencing `stream` so many times for the same values.